pdg/src/info.rs

```diff
@@ -1,31 +1,83 @@
 use crate::Graphs;
+use crate::graph::{Graph,NodeId,Node,NodeKind};
 use std::fmt::{self, Debug, Display, Formatter};
+use std::collections::HashMap;
 
 /// Information generated from the PDG proper that is queried by static analysis.
 ///
-/// Eventually this will include information about what kinds of [`Node`]s the [`Node`] is an ancestor of,
-/// as well as its ability to be used as a `&mut`.
+/// Includes information about what kinds of [`Node`]s the [`Node`] flows to,
+/// and eventually will also include its ability to be used as a `&mut`.
 ///
 /// [`Node`]: crate::graph::Node
-#[derive(Hash, Clone, PartialEq, Eq, Debug)]
-pub struct NodeInfo {}
+#[derive(Hash, Clone, PartialEq, Debug)]
+pub struct NodeInfo {
+    flows_to: Flows,
+}
+
+/// Contains information about what kinds of [`Node`]s the [`Node`] flows to.
+/// Load and store kinds contain both 
+///
+/// [`Node`]: crate::graph::Node.
+#[derive(Debug, Hash, Clone, Copy,PartialEq,Default)]
+pub struct Flows {
+    load: Option<NodeId>,
+    store: Option<NodeId>,
+    pos_offset: Option<NodeId>,
+    neg_offset: Option<NodeId>,
+}
 
 impl Display for NodeInfo {
     fn fmt(&self, f: &mut Formatter) -> fmt::Result {
         write!(f, "")
     }
 }
 
+fn init_flows(n_id: NodeId, n: &Node) -> Flows {
+    Flows {
+        load: matches!(n.kind, NodeKind::LoadAddr | NodeKind::LoadValue).then(|| n_id),
+        store: matches!(n.kind, NodeKind::StoreAddr | NodeKind::StoreValue).then(|| n_id),
+        pos_offset: matches!(n.kind, NodeKind::Offset(x) if x > 0).then(|| n_id),
+        neg_offset: matches!(n.kind, NodeKind::Offset(x) if x < 0).then(|| n_id),
+    }
+}
+
+
+/// Gathers information from a [`Graph`] (assumed to be acyclic and topologically sorted but not
+/// necessarily connected) for each [`Node`] in it whether there is a path following 'source' edges
+/// from any [`Node`] with a given property to the [`Node`] in question.
+///
+/// [`Node`]: crate::graph::Node
+/// [`Graph`]: graph::graph::Graph
+fn create_flow_info(g: &Graph) -> HashMap<NodeId, Flows> {
+    let mut f = HashMap::from_iter(
+        g.nodes
+            .iter_enumerated()
+            .map(|(idx, node)| (idx, init_flows(idx, node))),
+    );
+    for (n_id, node) in g.nodes.iter_enumerated().rev() {
+        let cur: Flows = *(f.get(&n_id).unwrap());
```

Comment:
```suggestion
        let node_flow_info: Flows = *(flow_map.get(&n_id).unwrap());
```

Reply:
It's named cur because there's both its and its parent's node_info being manipulated in the loop. Can I rename it to cur_node_flow_info instead?

Reply:
sure